stat/distmat/permutation.go

```diff
@@ -0,0 +1,53 @@
+// Copyright ©2020 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package distmat
+
+import (
+	"golang.org/x/exp/rand"
+
+	"gonum.org/v1/gonum/mat"
+)
+
+// UniformPermutation is a distribution over permutation matrices
+type UniformPermutation struct {
+	src rand.Source
+}
+
+// NewUniformPermutation constructs a new permutation matrix
+// generator using the given random source.
+func NewUniformPermutation(src rand.Source) *UniformPermutation {
+	up := UniformPermutation{src: src}
+	return &up
+}
+
+// Matrix draws a random permutuation matrix of dimension n.
+func (up *UniformPermutation) Matrix(n int) *mat.Dense {
+	m := mat.NewDense(n, n, nil)
+	up.MatrixTo(m)
+	return m
+}
+
+// MatrixTo sets the given matrix to be a random permutation matrix
+// panics if m is not square
+// Note this does not zero the matrix -- before calling this ensure it
+// is full of 0s.
+func (up *UniformPermutation) MatrixTo(m *mat.Dense) {
```

Comment:
We have another matrix distribution in this package, Wishart, in which the function sampling a random matrix is called RandSymTo. Neither this name or `MatrixTo` seems very fitting to me. We should decide on a consistent API now that we have more than one distribution in the package.

@kortschak ?

Reply:
Agreed. I was thinking the same thing last night. In other places we use `{noun,verb}To(dst *T)` for things like this. Figuring out what the correct noun or verb is is not necessarily easy. I don't think `Matrix` is the right noun.

Reply:
RandDenseTo would be consistent with wishart.  uniformPerm.RandDenseTo is a bit verbose though.

DenseTo() would go with SymTo() , CholTo() etc.  a distribution is surely spitting out random matrices so i don't think that is terribly unclear.

Reply:
FWIW, `uniformPerm.RandDenseTo(dst)` is not likely to exist in code anyway, you would be more likely to see `p.RandDenseTo(dst)`.

If you take a look at the factorisation types, they have things like `LTo`, `QTo`, etc. This reflects the semantics of the method rather than the type (this is also true for `SymTo` and `CholTo`, though the similarities between the semantics and the syntax confuse this). This is the basis of calling it `PermTo`.

Reply:
fair enough, i guess that is somewhat the price for not having overloading.  adjusted to PermTo.

Reply:
LGTM, thanks.